src/tab-group.ts

```diff
@@ -311,49 +328,70 @@ export default class TabGroup extends LitElement {
-    }
-  }
-
-  #onTabListResize() {
+  #onTabContainerResize() {
     if (this.#resizeTimeout) {
       clearTimeout(this.#resizeTimeout);
     }
 
-    // TODO
-    //
-    // This only needs to be called here so the indicator is updated when the content
-    // of Tab's slots changes.
-    //
-    // Tab's default slot will soon be replaced by a `label` attribute. When that
-    // happens, this call can be removed, and `#updateSelectedTabIndicator()` can be
-    // instead called when Tab dispatches "private-label-change" and
-    // "private-icon-slotchange" events.
-    //
-    // Those changes will certainly require slightly more code. But they'll make it
-    // much clearer why `#updateSelectedTabIndicator()` is being called. Additionally,
-    // Tab Group will no longer be doing unncessary work every time the viewport is
-    // resized.
-    this.#updateSelectedTabIndicator();
-
     // Toggling the overflow buttons will itself cause a resize. So we
     // wait a tick to avoid a loop.
     this.#resizeTimeout = setTimeout(() => {
       this.#setOverflowButtonsState();
     });
   }
 
+  #onTabDeselected() {
+    if (this.#firstTab && !this.#lastSelectedTab) {
+      this.#firstTab.selected = true;
+      this.#firstTab.tabIndex = 0;
+
+      this.#updateSelectedTabIndicator();
+    }
+
+    for (const panel of this.#panelElements) {
+      panel.privateSelected = panel.name === this.#lastSelectedTab?.panel;
+      panel.tabIndex = panel.name === this.#lastSelectedTab?.panel ? 0 : -1;
+    }
+  }
+
+  #onTabIconSlotOrLabelChange() {
+    if (this.#componentElementRef.value) {
+      // By temporarily disabling transitions, we ensure the measurements that happen in
+      // `#setOverflowButtonsState()` are in a settled, non-transitioning state.
+      // Otherwise, the overflow buttons would appear unnecessarily.
+      //
+      // Disabling these transitions also ensures the selected tab indicator and tab
+      // labels don't jump around briefly when being modified.
+      this.#componentElementRef.value.style =
+        '--private-transition-duration: 0ms';
+
+      this.#updateSelectedTabIndicator();
+
+      // Requesting an animation frame to help with the issue described above. We should
+      // wait until everything has settled completely before evaluating if the overflow
+      // buttons should appear or not. Otherwise they may appear and then immediately
+      // disappear.
+      requestAnimationFrame(() => {
+        this.#setOverflowButtonsState();
+
+        if (this.#componentElementRef.value) {
+          this.#componentElementRef.value.style = '';
+        }
+      });
+    }
+  }
+
+  #onTabListFocusOut() {
+    // Set the last selected tab as tabbable so that when pressing Shift + Tab on the
+    // Tab Panel focus goes back to the last selected tab.
+    for (const tabElement of this.#tabElements.values()) {
+      tabElement.tabIndex = tabElement === this.#lastSelectedTab ? 0 : -1;
+    }
+  }
```

Comment:
Renamed from `#onTabListFocusout()`.

The for loop used to read `for (const [, tabElement] of this.#tabElements.entries()) {`, but `entries()` isn't needed because we only wanted the tabElement. Clearer to use `.values()`.